src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/oops/ConstantPool.java

```diff
@@ -464,16 +467,35 @@ public int getMethodTypeIndexAt(int i) {
     return res;
   }
 
-  /** Lookup for multi-operand (InvokeDynamic, Dynamic) entries. */
-  public short[] getBootstrapSpecifierAt(int i) {
+  public int getBootstrapMethodsCount() {
+    U2Array operands = getOperands();
+    int count = 0;
+    if (operands != null) {
+      count = getOperandOffsetAt(0) / 2;
+    }
+    if (DEBUG) {
+      System.err.println("ConstantPool.getBootstrapMethodsCount: count = " + count);
+    }
+    return count;
+  }
+
+  public int getBootstrapMethodArgsCount(int bsmIndex) {
+    U2Array operands = getOperands();
     if (Assert.ASSERTS_ENABLED) {
-      Assert.that(getTagAt(i).isInvokeDynamic() || getTagAt(i).isDynamicConstant(), "Corrupted constant pool");
+      Assert.that(operands != null, "Operands is not present");
     }
-    int bsmSpec = extractLowShortFromInt(this.getIntAt(i));
+    int bsmOffset = getOperandOffsetAt(bsmIndex);
+    int argc = operands.at(bsmOffset + INDY_ARGC_OFFSET);
+    if (DEBUG) {
+      System.err.println("ConstantPool.getBootstrapMethodArgsCount: bsm index = " + bsmIndex + ", args count = " + argc);
+    }
+    return argc;
+  }
+
+  public short[] getBootstrapMethodAt(int bsmIndex) {
     U2Array operands = getOperands();
     if (operands == null)  return null;  // safety first
-    int basePos = VM.getVM().buildIntFromShorts(operands.at(bsmSpec * 2 + 0),
-                                                operands.at(bsmSpec * 2 + 1));
+    int basePos = getOperandOffsetAt(bsmIndex);
```

Comment:
Maybe you should pass `operands` into `getOperandOffsetAt()` so it does not need to be fetched again.

Reply:
Done.